lib/logger.js

```diff
@@ -173,7 +173,9 @@ Logger.prototype.log = function(statement, opts) {
     };
 
     if (opts) {
-        if (typeof opts === 'string') {
+        if (typeof opts === 'function') {
```

Comment:
We seem to do this a lot and it seems to me like a terrible idea - it's confusing to me that we can let people send in a callback function as the `opts` param. Wouldn't it be preferable to require them to send in an empty/null `opts` and be explicit about what `callback` is than this?

Reply:
Didn't answer my question, but I really dislike this - let's keep each parameter as it's meant to be instead of pseudo-overloading.

Reply:
sorry I did not see your comments.

Reply:
@dchai76 I think it's arguable. On one hand the implementation of  the code library becomes harder to read and maintain. On the other hand, usage of our code libraries becomes easier (you don't have to write your code likes this ``` logger.log("my lolg", {}, (err) => {if err stdout)}); ```  every time when you need to pass the cb to the log function. I think it's better to take the burden of readability on us than pass it on users.

Reply:
and besides, we can also pass a level to that. And it can grow into `log("logline", "", {}, cb)`. Which I think is not a really good practice.

Reply:
To be clear, I think this is bad for callers, not for our own maintainability - magic overloading can mask errors on their part. Did they mean to forget to include opts? Or was it intentional? The type of magic where we say one parameter can be treated in different ways is surprising, and not in a good way.

Reply:
I personally don't like this pattern, but I do see this used in lots of places :( 
I guess the best way to escape is the make it an async function that returns a promise, so up to the caller whether to chain it or not

Reply:
@weizou19 pointed out to me that Node itself does this (magic overloading of provided callbacks) (https://github.com/nodejs/node/blob/master/lib/fs.js#L405-L414) so as much as I dislike it I'm going to say I'm wrong on this.